osu.Game/Overlays/BeatmapListingOverlay.cs

```diff
@@ -260,10 +314,29 @@ private void addContentToPlaceholder(Drawable content)
             currentContent.FadeIn(200, Easing.OutQuint);
         }
 
+        private void addPanelsToContent(IEnumerable<DirectPanel> panels)
+        {
+            // TODO: Fade in?
+
+            if (currentContent == null)
+                return;
+
+            ((FillFlowContainer<DirectPanel>)currentContent).AddRange(panels);
```

Comment:
I don't really like how this requires casting. Is there a reason that the container is referred to as a `Drawable` rather than specifically as a `FillFlowContainer`? This seemed like some kind of future-proofing thing to me.

Reply:
The reason is this line, which makes it possible for `currentContent` to be a `NotFoundOverlay` (displayed when there are no search results):

https://github.com/ppy/osu/blob/9da76e3f3228aa6897d3b3aea31765ecce2724d5/osu.Game/Overlays/BeatmapListingOverlay.cs#L270

I wouldn't exactly call that ideal, though. They should probably not be clumped up in one field like that.

Reply:
Do you have a recommendation for separating these? I don't have quite enough experience with the codebase to know what an idiomatic way to handle this would be.

Reply:
I'd probably go for something like

```csharp
class BeatmapListingResultDisplay
{
    private NotFoundDrawable noResults; // probably needs renaming
    private FillFlowContainer<DirectPanel> resultPanels;

    public BindableList<BeatmapSetInfo> SearchResults { get; private set; }

    // subscribe to SearchResults.CollectionChanged
    // if eventArgs.Action == Add, async load panels and add to fill flow
    // if eventArgs.Action == Reset, display the no results drawable
    // only mutate SearchResults when a request has actually arrived
    // (to avoid mistakenly showing the no results drawable)
}
```

Has the upside of actually being testable without offline.